src/command.rs

```diff
@@ -271,226 +305,351 @@ pub fn parse(input: &str) -> Option<Vec<Command>> {
+                })?;
+                Command::Seek(seek_direction)
+            }
+            "focus" => {
+                let &target = args.get(0).ok_or(InsufficientArgs {
+                    cmd: command.into(),
+                    hint: Some("queue|search|library".into()),
+                })?;
+                // TODO: this really should be strongly typed
+                Command::Focus(target.into())
+            }
+            "save" => match args.get(0).cloned() {
+                Some("queue") => Ok(Command::SaveQueue),
+                Some(arg) => Err(BadEnumArg {
+                    arg: arg.into(),
+                    accept: vec!["**omit**".into(), "queue".into()],
+                }),
+                None => Ok(Command::Save),
+            }?,
+            "volup" => {
+                let amount = match args.get(0) {
+                    Some(&amount_raw) => {
+                        amount_raw.parse::<u16>().map_err(|err| ArgParseError {
+                            arg: amount_raw.into(),
+                            err: err.to_string(),
+                        })?
+                    }
+                    None => 1,
+                };
+                Command::VolumeUp(amount)
+            }
+            "voldown" => {
+                let amount = match args.get(0) {
+                    Some(&amount_raw) => {
+                        amount_raw.parse::<u16>().map_err(|err| ArgParseError {
+                            arg: amount_raw.into(),
+                            err: err.to_string(),
+                        })?
+                    }
+                    None => 1,
+                };
+                Command::VolumeDown(amount)
             }
-            "focus" => args
-                .get(0)
-                .map(|target| Command::Focus((*target).to_string())),
-            "save" => args
-                .get(0)
-                .map(|target| match *target {
-                    "queue" => Command::SaveQueue,
-                    _ => Command::Save,
-                })
-                .or(Some(Command::Save)),
-            "volup" => Some(Command::VolumeUp(
-                args.get(0).and_then(|v| v.parse::<u16>().ok()).unwrap_or(1),
-            )),
-            "voldown" => Some(Command::VolumeDown(
-                args.get(0).and_then(|v| v.parse::<u16>().ok()).unwrap_or(1),
-            )),
-            "help" => Some(Command::Help),
-            "reload" => Some(Command::ReloadConfig),
+            "help" => Command::Help,
+            "reload" => Command::ReloadConfig,
             "insert" => {
-                if args.is_empty() {
-                    Some(Command::Insert(None))
-                } else {
-                    args.get(0)
-                        .map(|url| Command::Insert(Some((*url).to_string())))
-                }
+                // IDEA: this should fail fast too
+                Command::Insert(args.get(0).map(|&url| url.into()))
```

Comment:
Currently if the user inputs a bad URL, the command will run and fail silently without feedback. Ideally I would like to make it fail fast here too, but I know too little about the command executor to make edits confidently.

Reply:
Ideally yes, we could at least validate the URL or check if it can be parsed using `SpotifyUrl::from_url`. Everything after that requires API calls.